tythonco/actions-sfdx
# CICD Enhancements

`To Test`

- [x] Clone the [project template repo](https://github.com/tythonco/sfdx-project-template)
- [x] Run through `npm run setup` and authenticate against a personal DevHub org, when asked state that DevHub = Prod
- [x] For testing purposes, say no to authenticating to a QA org but clone your `sfdx_auth_url.txt.enc` file into `qa_auth_url.txt.enc`
- [x] Ensure that you also have a `prod_auth_url.txt.enc` file that dupes `sfdx_auth_url.txt.enc`
- [x] Create a throwaway repo on GitHub and push your local source
- [x] Go to `Settings > Secrets > New repository secret` and give the new secret a name of `AUTH_FILE_KEY` and a value of the encryption key you used when going through `npm run setup`
- [x] Create a `.github` directory in your local repo with a subfolder of `workflows` that contains the four files below (note the reference to the cicd-enhancements branch of the repo in `uses` - eventually this will be @master instead):

`prod-deploy.yml`
```
name: Production Deployment Workflow
on:
  push:
    branches:
      - master
  workflow_dispatch:
jobs:
  prod-deployment:
    name: Production Deployment
    runs-on: ubuntu-latest
    steps:
    - name: Check out repository
      uses: actions/checkout@v2
    - name: Deploy to Production
      uses: tythonco/actions-sfdx@cicd-enhancements
      with:
        auth_file_key: ${{ secrets.AUTH_FILE_KEY }}
        command: deploy
        enc_auth_file: prod_auth_url.txt.enc
        test_level: RunLocalTests
```

`qa-deploy.yml`
```
name: QA Deployment Workflow
on:
  push:
    branches:
      - dev
  workflow_dispatch:
jobs:
  qa-deployment:
    name: QA Deployment
    runs-on: ubuntu-latest
    steps:
    - name: Check out repository
      uses: actions/checkout@v2
    - name: Deploy to QA
      uses: tythonco/actions-sfdx@cicd-enhancements
      with:
        auth_file_key: ${{ secrets.AUTH_FILE_KEY }}
        command: deploy
        enc_auth_file: qa_auth_url.txt.enc
```

`test-scratch.yml`
```
name: Testing Workflow
on:
  push:
  workflow_dispatch:
jobs:
  test:
    name: Testing
    runs-on: ubuntu-latest
    steps:
    - name: Check out repository
      uses: actions/checkout@v2
    - name: Test on Scratch Org
      uses: tythonco/actions-sfdx@cicd-enhancements
      with:
        auth_file_key: ${{ secrets.AUTH_FILE_KEY }}
        command: test-scratch
```

`validate.yml`
```
name: Validation Workflow
on:
  pull_request:
    branches:
      - dev
  workflow_dispatch:
jobs:
  validation:
    name: Validation
    runs-on: ubuntu-latest
    steps:
    - name: Check out repository
      uses: actions/checkout@v2
    - name: Validate via check-only deployment
      uses: tythonco/actions-sfdx@cicd-enhancements
      with:
        auth_file_key: ${{ secrets.AUTH_FILE_KEY }}
        command: validate
        enc_auth_file: prod_auth_url.txt.enc
```

- [x] Push your source up to GitHub
- [x] Ensure there is a `dev` branch in line with `master` on GitHub
- [x] Make a local change to the hello LWC and commit to git in a feature branch
- [x] Push the feature branch to GitHub
- [x] Ensure that the `Testing Workflow` action executes
- [x] Open a PR with `dev` as the base branch
- [x] Ensure that the `Validation Workflow` action executes
- [x] Ensure that the `QA Deployment Workflow` action executes
- [x] Open a PR with `master` as the base branch and merge
- [x] Ensure that the `Production Deployment Workflow` action executes
